artifactory/commands/golang/publish.go

```diff
@@ -0,0 +1,225 @@
+	goutils "github.com/jfrog/jfrog-cli-core/v2/utils/golang"
+	"github.com/jfrog/jfrog-client-go/artifactory"
+	_go "github.com/jfrog/jfrog-client-go/artifactory/services/go"
+	servicesutils "github.com/jfrog/jfrog-client-go/artifactory/services/utils"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils/checksum"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/utils/version"
+)
+
+// Publish go project to Artifactory.
+func publishPackage(packageVersion, targetRepo, buildName, buildNumber, projectKey string, servicesManager artifactory.ArtifactoryServicesManager) (*servicesutils.OperationSummary, []buildinfo.Artifact, error) {
+	projectPath, err := goutils.GetProjectRoot()
+	if err != nil {
+		return nil, nil, errorutils.CheckError(err)
+	}
+
+	collectBuildInfo := len(buildName) > 0 && len(buildNumber) > 0
+	modContent, modArtifact, err := readModFile(packageVersion, projectPath, collectBuildInfo)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Read module name
+	moduleName, err := goutils.GetModuleName(projectPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Info("Publishing", moduleName, "to", targetRepo)
+
+	props, err := utils.CreateBuildProperties(buildName, buildNumber, projectKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Temp directory for the project archive.
+	// The directory will be deleted at the end.
+	tempDirPath, err := fileutils.CreateTempDir()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer fileutils.RemoveTempDir(tempDirPath)
+
+	var zipArtifact *buildinfo.Artifact
+	params := _go.NewGoParams()
+	params.Version = packageVersion
+	params.Props = props
+	params.TargetRepo = targetRepo
+	params.ModuleId = moduleName
+	params.ModContent = modContent
+	params.ModPath = filepath.Join(projectPath, "go.mod")
+	params.ZipPath, zipArtifact, err = archive(moduleName, packageVersion, projectPath, tempDirPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	var artifacts []buildinfo.Artifact
+	if collectBuildInfo {
+		artifacts = []buildinfo.Artifact{*modArtifact, *zipArtifact}
+	}
+
+	// Create the info file if Artifactory version is 6.10.0 and above.
+	artifactoryVersion, err := servicesManager.GetConfig().GetServiceDetails().GetVersion()
+	if err != nil {
+		return nil, nil, err
+	}
+	version := version.NewVersion(artifactoryVersion)
+	if version.AtLeast(_go.ArtifactoryMinSupportedVersion) {
+		log.Debug("Creating info file", projectPath)
+		pathToInfo, err := createInfoFile(packageVersion)
+		if err != nil {
+			return nil, nil, err
+		}
+		defer os.Remove(pathToInfo)
+		if collectBuildInfo {
+			infoArtifact, err := createInfoFileArtifact(pathToInfo, packageVersion)
+			if err != nil {
+				return nil, nil, err
+			}
+			artifacts = append(artifacts, *infoArtifact)
+		}
+		params.InfoPath = pathToInfo
+	}
```

Comment:
It's going to be complicated, because it will cause passing many parameters to the new function. Also, most of the code of creating the info file is already in another function.